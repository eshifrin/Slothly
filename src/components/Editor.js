import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  TextInput,
  TouchableHighlight,
  Keyboard,
  ScrollView,
  Animated
} from 'react-native';
import DatePicker from 'react-native-datepicker';
import { formatDate, toJSDate } from '../utils/calcStats';
import { CancelIcon, UpdateIcon } from '../utils/Icons';
import PropTypes from 'prop-types';
import RenderIf from '../utils/RenderIf';

const Label = ({ text }) => (
  <View style={styles.label}>
    <Text style={styles.labelText}>{text}</Text>
  </View>
);

export default class Editor extends Component {
  constructor(props) {
    super();
    const { sessionLength = '', when = '', note = '' } = props.sessionToEdit;
    this.state = {
      sessionLength: String(sessionLength),
      when: formatDate(when),
      note
    };
  }

  componentDidMount = () => {
    this.setState({ basis: new Animated.Value(500) });
  };

  updateSession = () => {
    this.props.updateSession({
      when: toJSDate(this.state.when),
      sessionLength: this.state.sessionLength,
      note: this.state.note,
      id: this.props.sessionToEdit.id
    });
  };

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.buttons}>
          <TouchableHighlight
            onPress={this.props.onCancel}
            style={styles.button}
            underlayColor="white"
          >
            {CancelIcon}
          </TouchableHighlight>
          <TouchableHighlight
            onPress={this.updateSession}
            style={styles.button}
            underlayColor="white"
          >
            {UpdateIcon}
          </TouchableHighlight>
        </View>
        <Animated.View style={[{ maxHeight: this.state.basis }]}>
          <Label text={'SESSION LENGTH'} />

          <View style={styles.session}>
            <TextInput
              textAlign="right"
              style={[styles.sessionItem, styles.sessionLength]}
              keyboardType="number-pad"
              returnKeyType="done"
              value={this.state.sessionLength}
              onChangeText={sessionLength => this.setState({ sessionLength })}
            />
            <Text style={styles.sessionItem}>Minutes</Text>
          </View>
          <Label text="SESSION DATE" />
          <DatePicker
            style={[
              styles.session,
              { width: '100%', justifyContent: 'center', borderWidth: 0 }
            ]}
            date={this.state.when}
            mode="datetime"
            placeholder="select date"
            format="ddd h:mma | MMM Do, YYYY"
            maxDate={formatDate()}
            confirmBtnText="Confirm"
            cancelBtnText="Cancel"
            customStyles={{
              dateInput: {
                borderWidth: 0
              },
              dateText: {
                color: '#efdbc2',
                fontFamily: 'Futura',
                fontSize: 22,
                borderWidth: 0
              },
              dateTouchBody: {
                width: '100%',
                flexDirection: 'row-reverse'
              }
            }}
            onDateChange={when => {
              this.setState({ when });
            }}
          />
        </Animated.View>

        <Label text={'SESSION NOTES'} />

        <ScrollView style={styles.notes} keyboardDismissMode="interactive">
          <TextInput
            placeholderTextColor="#efdbc2"
            multiline={true}
            style={styles.notesnote}
            placeholder="  ... "
            value={this.state.note}
            onChangeText={note => this.setState({ note })}
            onBlur={this.moveNotesInputDown}
            onFocus={this.moveNotesInputUp}
            returnKeyType={'default'}
          />
        </ScrollView>
      </View>
    );
  }

  moveNotesInputUp = () => {
    Animated.timing(this.state.basis, {
      toValue: 0,
      duration: 300
    }).start();
  };

  moveNotesInputDown = () => {
    Animated.timing(this.state.basis, {
      toValue: 500,
      duration: 500
    }).start();
  };
}

Editor.propTypes = {
  sessionToEdit: PropTypes.shape({
    id: PropTypes.string,
    sessionLength: PropTypes.oneOf(PropTypes.string, PropTypes.number),
    when: PropTypes.instanceOf(Date),
    note: PropTypes.string
  }),
  updateSession: PropTypes.func.isRequired,
  onCancel: PropTypes.func
};

const styles = StyleSheet.create({
  container: {
    paddingTop: 30,
    flexDirection: 'column',
    backgroundColor: 'white',
    flex: 1
  },
  session: {
    flex: 0,
    flexDirection: 'row',
    alignItems: 'center',
    height: 75,
    backgroundColor: '#032b82'
  },
  sessionItem: {
    flex: 0,
    flexBasis: '50%',
    color: '#efdbc2',
    fontSize: 22,
    fontFamily: 'Futura',
    borderWidth: 0
  },
  sessionLength: {
    flexBasis: '35%',
    marginRight: 5
  },
  notes: {
    flex: 1
  },
  notesnote: {
    flex: 1,
    backgroundColor: '#032b82',
    color: '#efdbc2',
    fontSize: 22,
    fontFamily: 'futura',
    height: 300,
    maxHeight: 1000
  },
  buttons: {
    height: 80,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around',
    flex: 0
  },
  button: {
    flexDirection: 'row',
    justifyContent: 'center'
  },
  label: {
    height: 40,
    width: '100%',
    justifyContent: 'flex-end',
    borderBottomColor: 'black',
    borderBottomWidth: 0.5,
    paddingBottom: 5
  },
  labelText: {
    color: 'black',
    fontFamily: 'Futura',
    fontWeight: '700',
    fontSize: 14
  }
});
