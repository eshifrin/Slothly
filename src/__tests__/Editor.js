import 'react-native';
import React from 'react';
import Editor from '../components/Editor';
import { shallow } from 'enzyme';

describe('Editor', () => {
  let wrapper;
  const mockSession = {
    id: 'mockId1',
    sessionLength: '5',
    when: new Date(2018, 0, 1),
    note: 'mock note'
  };

  const mockUpdateSession = jest.fn();
  const mockOnCancel = jest.fn();

  beforeEach(done => {
    wrapper = shallow(
      <Editor
        sessionToEdit={mockSession}
        updateSession={mockUpdateSession}
        onCancel={mockUpdateSession}
      />
    );
    setTimeout(() => done(), 0);
  });

  it('renders', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls updateSession with correct parameters', () => {
    const newMockNote = 'new mock note';
    const newMockSession = { ...mockSession, ...{ note: newMockNote } };
    wrapper.setState(newMockSession);
    wrapper.instance().updateSession();
    expect(mockUpdateSession).toHaveBeenCalled;
    expect(mockUpdateSession).toHaveBeenCalledWith(newMockSession);
  });
});
