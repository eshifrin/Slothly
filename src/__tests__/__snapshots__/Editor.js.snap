// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor renders 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "column",
      "paddingTop": 30,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0,
        "flexDirection": "row",
        "height": 80,
        "justifyContent": "space-around",
      }
    }
  >
    <TouchableHighlight
      activeOpacity={0.85}
      onPress={[MockFunction]}
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
      underlayColor="white"
    >
      <Icon
        allowFontScaling={false}
        color="#efdbc2"
        name="circle-with-cross"
        size={75}
      />
    </TouchableHighlight>
    <TouchableHighlight
      activeOpacity={0.85}
      onPress={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
      underlayColor="white"
    >
      <Icon
        allowFontScaling={false}
        color="#032b82"
        name="ios-checkmark-circle-outline"
        size={75}
      />
    </TouchableHighlight>
  </View>
  <AnimatedComponent
    style={
      Array [
        Object {
          "maxHeight": 500,
        },
      ]
    }
  >
    <Label
      text="SESSION LENGTH"
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#032b82",
          "flex": 0,
          "flexDirection": "row",
          "height": 75,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        keyboardType="number-pad"
        onChangeText={[Function]}
        returnKeyType="done"
        style={
          Array [
            Object {
              "borderWidth": 0,
              "color": "#efdbc2",
              "flex": 0,
              "flexBasis": "50%",
              "fontFamily": "Futura",
              "fontSize": 22,
            },
            Object {
              "flexBasis": "35%",
              "marginRight": 5,
            },
          ]
        }
        textAlign="right"
        value="5"
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "borderWidth": 0,
            "color": "#efdbc2",
            "flex": 0,
            "flexBasis": "50%",
            "fontFamily": "Futura",
            "fontSize": 22,
          }
        }
      >
        Minutes
      </Text>
    </View>
    <Label
      text="SESSION DATE"
    />
    <DatePicker
      TouchableComponent={[Function]}
      androidMode="default"
      cancelBtnText="Cancel"
      confirmBtnText="Confirm"
      customStyles={
        Object {
          "dateInput": Object {
            "borderWidth": 0,
          },
          "dateText": Object {
            "borderWidth": 0,
            "color": "#efdbc2",
            "fontFamily": "Futura",
            "fontSize": 22,
          },
          "dateTouchBody": Object {
            "flexDirection": "row-reverse",
            "width": "100%",
          },
        }
      }
      date="Mon 12:00AM | Jan 1st, 2018"
      disabled={false}
      duration={300}
      format="ddd h:mma | MMM Do, YYYY"
      height={259}
      hideText={false}
      iconSource={
        Object {
          "testUri": "../../react-native-datepicker/date_icon.png",
        }
      }
      maxDate="Sun 11:16PM | Mar 18th, 2018"
      modalOnResponderTerminationRequest={[Function]}
      mode="datetime"
      onDateChange={[Function]}
      placeholder="select date"
      showIcon={true}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#032b82",
            "flex": 0,
            "flexDirection": "row",
            "height": 75,
          },
          Object {
            "borderWidth": 0,
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    />
  </AnimatedComponent>
  <Label
    text="SESSION NOTES"
  />
  <ScrollViewMock
    keyboardDismissMode="interactive"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      multiline={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="  ... "
      placeholderTextColor="#efdbc2"
      returnKeyType="default"
      style={
        Object {
          "backgroundColor": "#032b82",
          "color": "#efdbc2",
          "flex": 1,
          "fontFamily": "futura",
          "fontSize": 22,
          "height": 300,
          "maxHeight": 1000,
        }
      }
      value="mock note"
    />
  </ScrollViewMock>
</View>
`;
